resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/ExecutorPodFactory.scala

```diff
@@ -265,9 +268,53 @@ private[spark] class ExecutorPodFactoryImpl(
     val executorPodWithNodeAffinity =
         nodeAffinityExecutorPodModifier.addNodeAffinityAnnotationIfUseful(
             executorPodWithInitContainer, nodeToLocalTaskCount)
-    new PodBuilder(executorPodWithNodeAffinity)
+
+    val (executorPodWithTempLocalDirs, executorContainerWithTempLocalDirs) =
+        if (shuffleManager.isEmpty) {
```

Comment:
I wonder if we do not want this check. `spark.local.dir` is also used for block manager that can cache RDD output in local disks, whether or not external shuffle manager is used. From the documentation:

> Directory to use for "scratch" space in Spark, including map output files and RDDs that get stored on disk.

Reply:
The check needs to be a little more elaborate than that, I think.

The problem is that in `KubernetesExternalShuffleManager`, we have this block here:

```
  private val shuffleDirs = sparkConf.get(KUBERNETES_SHUFFLE_DIR).map {
    _.split(",")
  }.getOrElse(Utils.getConfiguredLocalDirs(sparkConf))
```

and then this:

```
  override def getExecutorShuffleDirVolumesWithMounts(): Seq[(Volume, VolumeMount)] = {
    shuffleDirs.zipWithIndex.map {
      case (shuffleDir, shuffleDirIndex) =>
        val volumeName = s"$shuffleDirIndex-${FilenameUtils.getBaseName(shuffleDir)}"
        val volume = new VolumeBuilder()
          .withName(volumeName)
          .withNewHostPath(shuffleDir)
          .build()
        val volumeMount = new VolumeMountBuilder()
          .withName(volumeName)
          .withMountPath(shuffleDir)
          .build()
        (volume, volumeMount)
    }
  }
}
```

Given these two blocks, sometimes the local dirs will be mounted as hostPath volumes. However, if the shuffle dirs are specified in the `KUBERNETES_SHUFFLE_DIR` Spark configuration, and these directories are not equal to what's set in `spark.local.dir`, then `spark.local.dir` will default to the Docker disk.

@foxish for thoughts - what's the purpose of having a separate configuration for shuffle directories specific to Kubernetes mode? That is - why do we expose `KUBERNETES_SHUFFLE_DIR` as a separate configuration option instead of falling back to mounting `apark.local.dir` as hostPath volumes?

Reply:
I see. Yes, I guess we could be better off by using spark.local.dir instead of KUBERNETES_SHUFFLE_DIR.

But if the user specifies spark.local.dir, should we interpret it as `hostPath` dirs in the cluster node as opposed to dirs in the container? I think that's a sensible interpretation for k8s, but wonder how other people think about it.

Reply:
There might be one benefit of having KUBERNETES_SHUFFLE_DIR as a separate option. If we use spark.local.dir, then the executor block managers may end up storing RDD caches in the same hostPath that the external shuffle service is overseeing. That could make block managers and external shuffle service step on each other's feet. Like deleting files in the wrong way as clean-up effort.

Reply:
I took a closer look at how things work here. These are the factors we need consider:

- When the external shuffle service is used, the executor advertises its `spark.local.dir`value (or the equivalent e.g. SPARK_LOCAL_DIR)  to the external shuffle service that is colocated on its node.
- The executor uses the `spark.local.dir` for both shuffle data and cache data regardless of if the external shuffle service is being used or not.
- The shuffle service must share the folders that the executor is using.

I thus propose the following:

- Remove `KUBERNETES_SHUFFLE_DIR` and always use `spark.local.dir` instead
- Mount paths from `spark.local.dir` as `hostPath` volumes if using the external shuffle service.
- Mount paths from `spark.local.dir` as `emptyDir` if not using the external shuffle service.

The downside here is that the directories used for the cache data will also be known and written to the same location as those for the shuffle data. However - I argue that this is already equal to the status quo, in that the shuffle service used in other cluster managers run into the same issues with e.g. disk space.

Reply:
In having the above semantics, the user is responsible for making sure that if they use the external shuffle service, that the directories set in `spark.local.dir` are always directories that are shared with the shuffle service that the executor runs on.

Reply:
I like the proposal. It's not as clean as we would want eventually, but it's a progress nonetheless from where we are. e.g. I can't find how KUBERNETES_SHUFFLE_DIR != spark.local.dir would work and maybe the proposal is fixing a bug in that edge case.

Reply:
I chatted with Matt offline and I'm +1 on doing the following in this PR:
- removing `KUBERNETES_SHUFFLE_DIR`
- putting spill / shuffle data onto a HostPath when there's an external shuffle service
- putting spill / shuffle data onto an EmptyDir when there's no external shuffle service
- leaving a TODO in the code to later move **spill** onto a EmptyDir volume while leaving **shuffle** on the HostDir (think this is better for perf, doing some direct disk benchmarks in k8s now)

Reply:
I gave some more thoughts on this. Logically there are two different work dirs, one for shuffle data and the other for spill data. So it'd be ideal to have two separate config keys.

In case of internal shuffle service is used, it's fine for the two to be same. For external shuffle service, the two are better to be separate. The spark core did not distinguish this and is causing us confusion. For instance, it's still not clear to me how to inform shuffle code to use a separate dir different from spark.local.dir.

I'm personally fine with a TODO addressing this need. @foxish, I wonder what you think?

Reply:
The TODO reminds us to revisit this when moving into upstream -- as we do that upstreaming I'd like to revisit this issue and potentially propose a change to core Spark to split shuffle data from spill data.

In k8s it's advantageous to have these on different disk types, since shuffle data should persist with the lifetime of the job (shuffle service), whereas spill data is not currently reusable through executor failures.  Shuffle should be on something more lasting (like a HostPath) and spill should be on something more ephemeral (like an EmptyDir).

Reply:
+1. Yes, I think the lifetime and ownership story is what matters. Thanks @mccheah and @ash211 for giving us chance to think about this.